tools/ecloud-revision-cleaner-tool/src/main/java/eu/europeana/cloud/jobs/RevisionRemoverJob.java

```diff
@@ -8,73 +8,154 @@
-        try {
-            int batchNumber = 1;
-            String startFrom = null;
-            do {
-                ResultSlice<CloudTagsResponse> resultSlice = dataSetServiceClient.getDataSetRevisionsChunk(revisionInformation.getProviderId(), revisionInformation.getDataSet(),
+
+        // Set up an executor service with the correct thread count.
+        ExecutorService executorService = Executors.newFixedThreadPool(this.threadCount);
+
+        // Loop through the cloud IDs, using result slicing (pagination).
+        String startFrom = null;
+        int totalCounter = 0;
+        do {
+
+            // Get the next result slice. Check that there is content.
+            ResultSlice<CloudTagsResponse> resultSlice = null;
+            try {
+                resultSlice = dataSetServiceClient.getDataSetRevisionsChunk(
+                        revisionInformation.getProviderId(),
+                        revisionInformation.getDataSet(),
                         revisionInformation.getRepresentationName(),
-                        revisionInformation.getRevisionName(), revisionInformation.getRevisionProvider(), revisionInformation.getRevisionTimeStamp(), startFrom, null);
+                        revisionInformation.getRevisionName(),
+                        revisionInformation.getRevisionProvider(),
+                        revisionInformation.getRevisionTimeStamp(), startFrom, null);
                 if (resultSlice == null || resultSlice.getResults() == null) {
-                    throw new DriverException("Getting cloud ids and revision tags: result chunk obtained but is empty.");
+                    LOGGER.error("Getting cloud IDs and revision tags: result chunk obtained but is empty.");
                 }
-                List<CloudTagsResponse> cloudTagsResponses = resultSlice.getResults();
-                for (CloudTagsResponse cloudTagsResponse : cloudTagsResponses) {
-                    List<Representation> representations = recordServiceClient.getRepresentationsByRevision(cloudTagsResponse.getCloudId(), revisionInformation.getRepresentationName(), revisionInformation.getRevisionName(), revisionInformation.getRevisionProvider(), revisionInformation.getRevisionTimeStamp());
-                    for (Representation representation : representations) {
-                        if (representation.getRevisions().size() == 1) {
-                            removeRepresentationWithFilesAndRevisions(cloudTagsResponse, representation);
-                            break;
-                        } else {
-                            removeRevisionsOnly(representation);
-                        }
-                    }
+            } catch (MCSException | RuntimeException e) {
+                logConnectionException(e);
+            }
+
+            // If we have not obtained any results, we are done.
+            if (resultSlice == null || resultSlice.getResults() == null) {
+                executorService.shutdownNow();
```

Comment:
Is possible scenario?
1. We get not null from result getDataSetRevisionsChunk
2. submit revisionRemove
3. go to next iteration of "do" loop.
4. getDataSetRevisionsChunk returns null because there is no more data
5. invoke ExecutorService.shutdownNow, 
6. Not all "revision removing" have been done already, and these tasks are queued in executor service, and would not be executed

Reply:
I think your point 4 is a false assumption: it wouldn't return null if there is no more data, but it would return an empty (not null) structure. And we would know because of the way startFrom is calculcated and checked. 

But if you're concerned, we can perhaps remove the shutDownNow(), and let it finish on it's own.